mlir/lib/Dialect/Rock/Transforms/ThreadwiseGemmLowering.cpp

```diff
@@ -89,6 +89,168 @@ struct ThreadwiseReadIntoRewritePattern
+
+      SmallVector<int64_t, 3> dimensions = {k, i, j, kPack};
+      SmallVector<int64_t, 3> strides = {1, 1, 1, 1};
+       
+      normalizedViewA = normalizeView(b, loc, {"k", "i", "j", "kpack"}, dimensions, true, {"j"});
+      normalizedViewB = normalizeView(b, loc, {"k", "i", "j", "kpack"}, dimensions, true, {"i"});
+      normalizedViewC = normalizeView(b, loc, {"k", "i", "j", "kpack"}, dimensions, true, {"k", "kpack"});
+    }
+
+    auto [rawBufferA, bufferViewA, sourceANeeds64BitIdx] =
+          untransform(b, bufferA, normalizedViewA);
+    auto [rawBufferB, bufferViewB, sourceBNeeds64BitIdx] =
+          untransform(b, bufferB, normalizedViewB);
+    auto [rawBufferC, bufferViewC, dstNeeds64BitIdx] =
+          untransform(b, bufferC, normalizedViewC);
+    assert(!sourceANeeds64BitIdx && "Registers shouldn't need 64-bit indexing");
+    assert(!sourceBNeeds64BitIdx && "Registers shouldn't need 64-bit indexing");
+    assert(!dstNeeds64BitIdx && "Registers shouldn't need 64-bit indexing");
+
+    // Loop properties
+    auto computeStart = llvm::to_vector(op.getComputeIndices());
+
+    if(isMfma || isWmma){
+
+      // Emit the loop
+      auto accelLoop = b.create<TransformingForOp>(
+          loc, ArrayRef<ValueRange>{computeStart, computeStart, computeStart},
+          ArrayRef<Attribute>{bufferViewA, bufferViewB, bufferViewC},
+          /*bounds=*/ArrayRef<int64_t>{1, 1, 1},
+          /*strides=*/ArrayRef<int64_t>{1, 1, 1},
+          /*forceUnroll=*/true, /*useIndexDiffs=*/true);
+      {
+        OpBuilder::InsertionGuard guard(b);
+        b.setInsertionPointToStart(accelLoop.getBody());
+        auto coordsA = accelLoop.getLowerCoords(/*domain=*/0);
+        auto coordsB = accelLoop.getLowerCoords(/*domain=*/1);
+        auto coordsC = accelLoop.getLowerCoords(/*domain=*/2);
+
+        Value argA = b.create<memref::LoadOp>(loc, argTypeA, rawBufferA, coordsA);
+        Value argB = b.create<memref::LoadOp>(loc, argTypeB, rawBufferB, coordsB);
+        // llvm::outs() << "\n*** argA: " << argA << "\n"
+        //              << "\n*** argB: " << argB << "\n";
+        emitter->emitThreadwiseLoop(b, loc, argA, argB, rawBufferC, coordsC);
+      }
+      b.eraseOp(op);
+      return success();
+    } else {
+      
+      int64_t k = bufferAShape.back();
+      int64_t i = *(bufferCShape.end() - 2);
+      int64_t kPack = bufferAShape[2];
+      int64_t j = bufferCShape.back();
+
+      SmallVector<int64_t, 3> dimensions = {k, i, j, kPack};
+
+      auto gemmLoop = b.create<TransformingForOp>(
+          loc, ArrayRef<ValueRange>{computeStart, computeStart, computeStart},
+          ArrayRef<Attribute>{bufferViewA, bufferViewB, bufferViewC}, dimensions,
+          /*strides=*/std::nullopt, /*forceUnroll=*/true,
+          /*useIndexDiffs=*/false);
+
+      {
+        OpBuilder::InsertionGuard guard(b);
+        b.setInsertionPointToStart(gemmLoop.getBody());
+
+        auto coordsA = gemmLoop.getLowerCoords(/*domain=*/0);
```

Comment:
Why do you need this case? What prevents you from using `emitAccelLoop` here?

Reply:
The problem is caused by different ways of creating gemmLoop and accelLoop TransformingForOp. I want to try to equalize these approaches by refactoring the blockwise level.